src/coreclr/vm/invokeutil.cpp

```diff
@@ -124,191 +124,119 @@ void *InvokeUtil::GetIntPtrValue(OBJECTREF pObj) {
-    else {
-        pMT = 0;
-        oType = ELEMENT_TYPE_OBJECT;
-    }
-    type = th.GetVerifierCorElementType();
-
-    // This basically maps the Signature type our type and calls the CreatePrimitiveValue
-    //  method.  We can omit this if we get alignment on these types.
     switch (type) {
     case ELEMENT_TYPE_BOOLEAN:
-    case ELEMENT_TYPE_I1:
     case ELEMENT_TYPE_U1:
+    case ELEMENT_TYPE_I1:
+    {
+        _ASSERTE(rArg != NULL);
+        *(INT8 *)pArgDst = *(INT8 *)rArg;
+        break;
+    }
+
     case ELEMENT_TYPE_I2:
     case ELEMENT_TYPE_U2:
     case ELEMENT_TYPE_CHAR:
+    {
+        _ASSERTE(rArg != NULL);
+        *(INT16 *)pArgDst = *(INT16 *)rArg;
+        break;
+    }
+
     case ELEMENT_TYPE_I4:
     case ELEMENT_TYPE_U4:
     case ELEMENT_TYPE_R4:
-    IN_TARGET_32BIT(case ELEMENT_TYPE_I:)
     IN_TARGET_32BIT(case ELEMENT_TYPE_U:)
+    IN_TARGET_32BIT(case ELEMENT_TYPE_I:)
     {
-        // If we got the univeral zero...Then assign it and exit.
-        if (rObj == 0)
-            *(PVOID *)pArgDst = 0;
-        else
-        {
-            ARG_SLOT slot;
-            CreatePrimitiveValue(type, oType, rObj, &slot);
-            *(PVOID *)pArgDst = (PVOID)slot;
-        }
+        _ASSERTE(rArg != NULL);
+        *(INT32 *)pArgDst = *(INT32 *)rArg;
         break;
     }
 
-
     case ELEMENT_TYPE_I8:
     case ELEMENT_TYPE_U8:
     case ELEMENT_TYPE_R8:
     IN_TARGET_64BIT(case ELEMENT_TYPE_I:)
     IN_TARGET_64BIT(case ELEMENT_TYPE_U:)
     {
-        // If we got the univeral zero...Then assign it and exit.
-        if (rObj == 0)
-            *(INT64 *)pArgDst = 0;
-        else
-        {
-            ARG_SLOT slot;
-            CreatePrimitiveValue(type, oType, rObj, &slot);
-            *(INT64 *)pArgDst = (INT64)slot;
-        }
+        _ASSERTE(rArg != NULL);
+        *(INT64 *)pArgDst = *(INT64 *)rArg;
         break;
     }
 
     case ELEMENT_TYPE_VALUETYPE:
     {
-        // If we got the universal zero...Then assign it and exit.
-        if (rObj == 0) {
-            InitValueClassArg(argDest, th.AsMethodTable());
-         }
-        else {
-            if (!th.AsMethodTable()->UnBoxIntoArg(argDest, rObj))
+        if (Nullable::IsNullableType(th))
+        {
+            // ASSUMPTION: we only receive T or NULL values, not Nullable<T> values
+            // and the values are boxed, unlike other value types.
+            MethodTable* pMT = th.AsMethodTable();
+            OBJECTREF src = (OBJECTREF)(Object*)*rArg;
+            if (!pMT->UnBoxIntoArg(argDest, src))
                 COMPlusThrow(kArgumentException, W("Arg_ObjObj"));
         }
+        else if (rArg == NULL) {
+            // A byref-like type can't be boxed but the Invoke() arg value can be passed as a null object where that is
+            // marshalled as a null reference to indicate we need to initialize the byref-like type here.
+            MethodTable* pMT = th.GetMethodTable();
+            if (pMT->IsByRefLike())
```

Comment:
That would require a breaking change - this PR doesn't change the support. The best time to do that is when we have byref working, so there would be an alternative. However, the workaround today if we add the breaking change now is to manually add a wrapper method that doesn't take the ByRefLike type which then calls the target method.

Reply:
> The best time to do that is when we have byref working, so there would be an alternative.

Not sure I agree with this. Whatever anyone is doing with the current support is likely wrong and providing support for a scenario that is fundamentally broken. This is likely a historical oversight that I'd like to understand earlier rather than later. I'd remove this soon so we can message that.

Reply:
I'll see if there's any impact from removing it besides arcane tests.

Reply:
A summary of current behavior (before this PR) also comparing with a value type:
- Value type:
  - If `null` is passed, a default value is supplied
  - The target parameter can be by-ref
  - A non-null value is supported
- ByRefLike type:
  - If `null` is passed, a default value is supplied **(we should probably just throw NSE)**
  - The target parameter **cannot** be by-ref (NSE is thrown)
  - A non-null value is **not** supported (compile error since ByRefLike can't be boxed)

Reply:
@GrabYourPitchforks can you comment on removing support for allowing `null` to be passed to methods taking ByRefLike types? It looks like you added some of the tests that need removing in https://github.com/dotnet/runtime/pull/40632.

From a recent discussion, this support is somewhat arcane and will be addressed properly when we support ByRefLike types directly with new Invoke APIs down the road.

It is getting in the way now since we are moving the native code to managed code and it's not possible to create a default ref struct and box it.

The workaround for anyone affected is to add a wrapper method that doesn't by the ByRefLike parameters, and then call the target method that has the ByRefLike parameters.

Reply:
The latest commit added the breaking change. There were only very specific tests for this - nothing related to product features for example.